Refactor SLRU download interface between Postgres and the neon extension

hlinnaka opened the pull request:
Move the responsibility of writing the SLRU segment contents to the neon extension. The neon extension hook now downloads the file and places it to the correct path, and returns a true/false to the caller to indicate if the file existed. On successful download, the caller retries opening the file.

MMeent commented on src/backend/access/transam/slru.c:
```diff
@@ -762,45 +764,7 @@ SimpleLruDownloadSegment(SlruCtl ctl, int pageno, char const* path)
-	}
-	pfree(buffer);
-	return fd;
+	return smgr_read_slru_segment(&dummy_smgr_rel, path, segno);
```
If we're changing this API, can we please remove the `slru` argument? It's a bit of a pain to construct this dummy SLRU, and we don't use it inside PG's nor Neon's code.

MMeent commented on src/include/storage/smgr.h:
```diff
@@ -131,7 +131,7 @@ typedef struct f_smgr
 	void		(*smgr_end_unlogged_build) (SMgrRelation reln);
 
-	int  		(*smgr_read_slru_segment) (SMgrRelation reln, const char *path, int segno, void* buffer);
+	bool  		(*smgr_read_slru_segment) (SMgrRelation reln, const char *path, int segno);
```
Let's make this a proper hook rather than a member of `f_smgr` - the SMgrRelation argument is unused in Neon, and the only relation between `smgr` and this is that they happen to read data. However, lacking buffered SLRUs, keeping this function in the smgr struct makes very little sense when `md.c` could be the authorative smgr for some relations (e.g. backend-local ones)

hlinnaka replied:
Ok, done! I copied how you did it on the REL_18_STABLE_neon branch

hlinnaka committed:
Refactor SLRU download interface between Postgres and the neon extension

Move the responsibility of writing the SLRU segment contents to the
neon extension. The neon extension hook now downloads the file and
places it to the correct path, and returns a true/false to the caller
to indicate if the file existed. On successful download, the caller
retries opening the file.

While we're at it, refactor the read_slru_segment function so that it
is not part of the SMGR api anymore, but an independent hook. We were
previously pretending it's part of the SMGR api and because of that,
passed a dummy SMgrRelation, but it was a bit silly.

MMeent reviewed (approved):
LGTM at first glance.
